CurvesAlgoUpdateEndpointMultiplicity : Correct Catmull-Rom

danieldresser-ie opened the pull request:
Catmull-Rom only needs double up endpoints, not tripled.

This also corrects handling of bezier curves while I'm at - bezier curves don't have any endpoint multiplicity.  If we want something for adding / removing tangents, that should probably be a separate function.

This fixes Carsten's overshoots, and a basic test passes.  The one more thing I should probably do is add tests that converting linear->bSpline->catmull yields the same result as linear->catmull.

If I was going to do a bit more playing with this while I'm at it, the one thing that might be worth checking into is why catmull-rom curves seem to always draw black in Gaffer ( whereas linear curves are white or use Cs ), and why gl:curvesPrimitive:ignoreBasis appears to default on?

andrewkaufman commented on src/IECoreScene/CurvesAlgoUpdateEndpointMultiplicity.cpp:
```diff
@@ -46,110 +46,74 @@ using namespace Imath;
-template<typename T>
-void expandVarying( const std::vector<T> &in, std::vector<T> &out, const CurvesPrimitive *curves, const PrimitiveVariable &primVar, const IECore::CubicBasisf &targetBasis )
+// Set the endpoint multiplicities for vertex and varying primvars, based on the given basis
+void setMultiplicity( const IECore::CubicBasisf &cubicBasis, int &vertexMultiplicity, int &varyingMultiplicity )
```
Or could it take both orig and new basis as args and return the final adjustment values instead? That'd reduce the call site from 6 lines down to 1.

danieldresser-ie replied:
It's a separate function to avoid having to duplicate the if/else for mapping basis -> endPointMultiplicity for both input and output.

Having two outputs was a bit ugly though.  I've switched the whole thing up to use CurvesPrimitive::numSegments to derive varyingAdjustment from vertexAdjustment instead of manually specifying the varying values.  See if this feels better to you.

andrewkaufman replied:
Thanks, I do find it easier to follow now

andrewkaufman commented on src/IECoreScene/CurvesAlgoUpdateEndpointMultiplicity.cpp:
```diff
@@ -46,110 +46,74 @@ using namespace Imath;
 struct DuplicateEndPoints
 {
-	DuplicateEndPoints( const IECore::CubicBasisf &targetBasis, bool expand ) : m_targetBasis( targetBasis ), m_expand( expand )
+	DuplicateEndPoints( const IECore::CubicBasisf &targetBasis, int adjustVertex, int adjustVarying ) : m_targetBasis( targetBasis ), m_adjustVertex( adjustVertex ), m_adjustVarying( adjustVarying )
```
similar to my suggest rename above, perhaps `vertexAdjustment` and `varyingAdjustment`?

andrewkaufman commented on src/IECoreScene/CurvesAlgoUpdateEndpointMultiplicity.cpp:
```diff
@@ -162,28 +126,10 @@ struct DuplicateEndPoints
-			}
-		}
+
+		int adjust = primVar.interpolation == PrimitiveVariable::Vertex ? m_adjustVertex : m_adjustVarying;
```
Isn't this going to attempt to adjust Uniform and Constant primvars as though they were Varying?

danieldresser-ie replied:
We only run EndpointDuplicator on Vertex and Varying primvars.  Do you want an assert or something?

andrewkaufman commented:
> The one more thing I should probably do is add tests that converting linear->bSpline->catmull yields the same result as linear->catmull.

I do think this additional test would be worthwhile

danieldresser-ie commented:
I think I've now addressed all feedback here.  Would be great if Andrew could review CurvesAlgoUpdateEndpointMultiplicity again, and either Andrew or John could take a look through the GL fixes I've snuck in here - the basic idea of the GL fix I've discussed with John, but I'm not totally confident on why we're wrapping some of this stuff into macros, so that work should definitely be checked.

andrewkaufman commented on src/IECoreScene/CurvesAlgoUpdateEndpointMultiplicity.cpp:
```diff
@@ -46,110 +46,74 @@ using namespace Imath;
+	out.reserve( in.size() + numCurves * 2 * adjust );
 
-	size_t index = 0;
+	size_t curveIndex = 0;
```
I think this could still do with a rename, unless I interpreted it wrong?

andrewkaufman commented:
Thanks Dan, I've looked over the algo changes I think think its easier to follow now. I've totally ignored the GL changes though, hopefully John can review those.

danieldresser-ie commented:
OK, I think I've addressed Andrew's last naming comment, and hopefully the tests pass now.

I've squashed all the endPointMult commits together ( since the fixes were spread across them ), and updated Changes ( are we supposed to be doing that in Cortex now? )

andrewkaufman commented:
> and updated Changes ( are we supposed to be doing that in Cortex now? )

We haven't taken a stance on it yet. FYI, I think technically the Gaffer workflow would've had your 2nd & 3rd commits each add one line to the Changes file.
